core/src/test/scala/org/apache/spark/deploy/history/EventLogFileWritersSuite.scala

```diff
@@ -0,0 +1,400 @@
+    buildWriterAndVerify(conf, classOf[RollingEventLogFilesWriter])
+
+    conf.set(EVENT_LOG_ENABLE_ROLLING, false)
+    buildWriterAndVerify(conf, classOf[SingleEventLogFileWriter])
+  }
+
+  val allCodecs = Seq(None) ++
+    CompressionCodec.ALL_COMPRESSION_CODECS.map(c => Some(CompressionCodec.getShortName(c)))
+
+  allCodecs.foreach { codecShortName =>
+    test(s"initialize, write, stop - with codec $codecShortName") {
+      val appId = getUniqueApplicationId
+      val attemptId = None
+
+      val conf = getLoggingConf(testDirPath, codecShortName)
+      val writer = createWriter(appId, attemptId, testDirPath.toUri, conf,
+        SparkHadoopUtil.get.newConfiguration(conf))
+
+      writer.start()
+
+      // snappy stream throws exception on empty stream, so we should provide some data to test.
+      val dummyData = Seq("dummy1", "dummy2", "dummy3")
+      dummyData.foreach(writer.writeEvent(_, flushLogger = true))
+
+      verifyWriteEventLogFile(appId, attemptId, testDirPath.toUri, codecShortName,
+        isCompleted = false, dummyData)
+
+      writer.stop()
+
+      verifyWriteEventLogFile(appId, attemptId, testDirPath.toUri, codecShortName,
+        isCompleted = true, dummyData)
+    }
+  }
+
+  test("spark.eventLog.compression.codec overrides spark.io.compression.codec") {
+    val conf = new SparkConf
+    conf.set(EVENT_LOG_COMPRESS, true)
+    val hadoopConf = SparkHadoopUtil.get.newConfiguration(conf)
+
+    val appId = "test"
+    val appAttemptId = None
+
+    // The default value is `spark.io.compression.codec`.
+    val writer = createWriter(appId, appAttemptId, testDirPath.toUri, conf, hadoopConf)
+    assert(writer.compressionCodecName.contains("lz4"))
+
+    // `spark.eventLog.compression.codec` overrides `spark.io.compression.codec`.
+    conf.set(EVENT_LOG_COMPRESSION_CODEC, "zstd")
+    val writer2 = createWriter(appId, appAttemptId, testDirPath.toUri, conf, hadoopConf)
+    assert(writer2.compressionCodecName.contains("zstd"))
+  }
+
+  protected def readLinesFromEventLogFile(log: Path, fs: FileSystem): List[String] = {
+    val logDataStream = EventLogFileReader.openEventLog(log, fs)
+    try {
+      Source.fromInputStream(logDataStream).getLines().toList
+    } finally {
+      logDataStream.close()
+    }
+  }
+
+  protected def createWriter(
+      appId: String,
+      appAttemptId : Option[String],
+      logBaseDir: URI,
+      sparkConf: SparkConf,
+      hadoopConf: Configuration): EventLogFileWriter
+
+  protected def verifyWriteEventLogFile(
+      appId: String,
+      appAttemptId : Option[String],
+      logBaseDir: URI,
+      compressionCodecShortName: Option[String],
+      isCompleted: Boolean,
+      expectedLines: Seq[String] = Seq.empty): Unit
+
+  protected def skipVerifyEventLogFile(
+      compressionCodecShortName: Option[String],
+      isCompleted: Boolean): Boolean = {
+    // Spark initializes LZ4BlockOutputStream with syncFlush=false, so we can't force
```

Comment:
This comment makes me think that trying to test the behavior of "what data is available in a file currently being written" is not a good test. Spark makes a best effort to get data available from the file when replaying, but past that, this test seems to be trying to enforce something that at times is out of Spark's control.

Reply:
I guess SHS is on the same page (I'm not sure it can be remedied with different JVM), so if there's some issue here, SHS may have same issue as well. Maybe we would be better to make problematic compressions as unsupported?

Reply:
> Maybe we would be better to make problematic compressions as unsupported?

That's not what I mean. The SHS does the best it can; what I'm saying is that these tests being added that try to read open files aren't really helpful since they depend on specific behavior of code that is outside of Spark's control - namely, compression codecs.

So the test really should only be asserting that data is available in the file after the file is closed; before that, there is no guarantee that data has actually made it to the file.

Reply:
> SHS does the best it can; 

OK, that clarifies everything. Then no strong opinion to keep the test on reading event log file before it's closed. Will remove.